android-components/components/service/contile/src/main/java/mozilla/components/service/contile/ContileTopSitesProvider.kt

```diff
@@ -117,51 +133,138 @@ class ContileTopSitesProvider(
-        diskCacheLastModified?.let { return it }
+    internal fun isCacheExpired(): Boolean {
+        cacheState.cacheMaxAge?.let { return Date().time > it }
 
         val file = getBaseCacheFile()
 
-        return if (file.exists()) {
-            file.lastModified().also {
-                diskCacheLastModified = it
+        cacheState =
+            if (file.exists()) {
+                cacheState.computeMaxAges(
+                    file.lastModified(),
+                    maxCacheAgeInSeconds * DateUtils.SECOND_IN_MILLIS,
+                    (readFromDiskCache()?.validFor ?: 0L) * DateUtils.SECOND_IN_MILLIS,
+                )
+            } else {
+                cacheState.invalidate()
             }
-        } else {
-            -1
-        }
+
+        // If cache is invalid, we should also consider it as expired
+        return Date().time > (cacheState.cacheMaxAge ?: -1L)
     }
 
     private fun getCacheFile(): AtomicFile = AtomicFile(getBaseCacheFile())
 
     @VisibleForTesting
     internal fun getBaseCacheFile(): File = File(applicationContext.filesDir, CACHE_FILE_NAME)
+
+    /**
+     * Data stored in the cache file
+     *
+     * @param validFor Time in milliseconds describing the click validity for the set of top sites.
+     * @param topSites List of provided top sites.
+     */
+    internal data class CachedData(
+        val validFor: Long,
+        val topSites: List<TopSite.Provided>,
+    )
+
+    /**
+     * Current state of the cache.
+     *
+     * @param shouldUseServerMaxAge Whether or not [serverCacheMaxAge] should be used instead of
+     * [localCacheMaxAge].
+     * @param isCacheValid Whether or not the current set of cached top sites is still valid.
+     * @param localCacheMaxAge Maximum unix timestamp until the current set of cached top sites
+     * is still valid, specified by the client.
+     * @param serverCacheMaxAge Maximum unix timestamp until the current set of cached top sites
+     * is still valid, specified by the server.
+     */
+    internal data class CacheState(
+        val shouldUseServerMaxAge: Boolean,
+        val isCacheValid: Boolean = true,
+        val localCacheMaxAge: Long? = null,
+        val serverCacheMaxAge: Long? = null,
+    ) {
+        val cacheMaxAge
+            get() = if (isCacheValid) {
+                if (shouldUseServerMaxAge) serverCacheMaxAge else localCacheMaxAge
+            } else {
+                null
+            }
+
+        fun invalidate(): CacheState =
+            this.copy(isCacheValid = false, localCacheMaxAge = null, serverCacheMaxAge = null)
+
+        /**
+         * Update local and server max age values
+         */
+        fun computeMaxAges(lastModified: Long, localMaxAge: Long, serverMaxAge: Long): CacheState =
+            this.copy(
+                isCacheValid = true,
+                localCacheMaxAge = lastModified + localMaxAge,
+                serverCacheMaxAge = lastModified + serverMaxAge,
+            )
+    }
```

Comment:
I think these types and comments add a lot of semantic value, thanks again for indulging me!